@constructedBy attribute

marek opened the pull request:
Added @constructedBy to document how objects are created that do not use the regular constructor.
For example:

HTMLAudioElement is created by new Audio()
HTMLVideoElement is created by document.createElement("video");

if a @constructedBy is specified, and no @constructor is present, the default constructor is not generated.
if both @constructedBy and @constructor are present. Both are generated

/**
- this is a MyClass description
- @class
- @constructor
- @param {Number} [foo] foo is a constructor param
- @constructedBy {document.createElement("MyClass")} {This is a description}
- // this is an example of how to create a new instance of MyClass
- var instance = document.createElement("MyClass");
  */

MyClass = function(foo) { }

jeffheifetz commented on build/bbPlugin.js:
```diff
@@ -33,7 +66,9 @@ JSDOC.PluginManager.registerPlugin(
-		var learnTag = symbol.comment.getTag("learns");
+                var learnTag = symbol.comment.getTag("learns");
+                var constructorTag = symbol.comment.getTag("constructor");
+                var constructedBy = symbol.comment.tags.filter(function($){return $.title=="constructedBy" && !!$.type}); 
```
Is that a double negative?

marek replied:
http://navirudra.com/blog/javascript/javascript-double-negative-trick.html

jeffheifetz replied:
Nice trick, Thanks! I never knew about this.
If it covers additional cases, should we throw this everywhere?

jeffheifetz commented on build/bbPlugin.js:
```diff
@@ -126,9 +169,18 @@ JSDOC.PluginManager.registerPlugin(
+                        LOG.warn(docTag.title + " missing type");
+                    } else {
+                        var parts = GetType(docTag.desc);
+                        docTag.desc = parts.type;
```
The tag should already be parsed. You should not need to do so again

marek replied:
This parses two sets. jsDocs only parses one by default.

jeffheifetz replied:
Sorry didn't notice

jeffheifetz commented on build/BBTemplate/class.tmpl:
```diff
@@ -436,6 +446,20 @@
 			</if>
 		</div>
 	</if>
+    {+ResetIndex()+}
```
I cannot make this call from BBPlugin. If you want to be able to you'll need to declare these in the publish.js

jeffheifetz replied:
Got it working with our help, thanks

jeffheifetz commented:
Seems like this should be good enough for now, we can always expand it in the future if we need different platform support per constructor

marek committed:
more comments

jeffheifetz commented:
Is there a way to set different sets of parameters for each constructor?

marek committed:
Added @constructedBy to document how objects are created that do not use the regular constructor.
For example:

HTMLAudioElement is created by new Audio()
HTMLVideoElement is created by document.createElement("video");

if a @constructedBy is specified, and no @constructor is present, the default constructor is not generated.
if both @constructedBy and @constructor are present. Both are generated

/**
* this is a MyClass description
* @class
* @constructor
* @param {Number} [foo] foo is a constructor param
* @constructedBy {document.createElement("MyClass")} {This is a description}
* // this is an example of how to create a new instance of MyClass
* var instance = document.createElement("MyClass");
*/

MyClass = function(foo) { }

Conflicts:

	build/bbPlugin.js

marek committed:
merge fixes

marek commented:
Nope there is no way of defining the parameters. 
Felt that it would be odd to describe the parameter types for a call like document.createElement("audio"); when it's a fixed string you use.

jeffheifetz commented:
How will Jack use it for something like window.openDatabase that has like 4 parameters?

----- Original Message -----
From: marek [mailto:reply@reply.github.com]
Sent: Tuesday, May 10, 2011 01:28 PM
To: Jeffrey Heifetz
Subject: Re: [WebWorks-API-Docs] @constructedBy attribute (#35)

Nope there is no way of defining the parameters. 
Felt that it would be odd to describe the parameter types for a call like document.createElement("audio"); when it's a fixed string you use.

## 

Reply to this email directly or view it on GitHub:
https://github.com/blackberry/WebWorks-API-Docs/pull/35#issuecomment-1132288

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

marek commented:
It's limited to just providing information in the description field. Or we can document 'window' more thoroughly
